src/main.rs

```diff
@@ -104,30 +115,151 @@ fn run<F>(future: F, threads_config: ThreadsConfig)
+                .fold(to_client, move |to_client, packet| {
+                    debug!("Send {:?} => {:?}", client_pk, packet);
+                    to_client.send(packet)
+                        .deadline(Instant::now() + Duration::from_secs(30))
+                        .map_err(|_|
+                            Error::new(ErrorKind::Other,
+                                format!("Writer timed out"))
+                        )
+                })
+                // drop to_client when to_client_rx stream is exhausted
+                .map(|_to_client| ())
+                .map_err(|_|
+                    Error::new(ErrorKind::Other,
+                        format!("Writer ended with error"))
+                );
+
+            // reader = for each Packet from client send it to server processor
+            let reader = from_client
+                .forward(from_client_tx
+                    .sink_map_err(|e|
+                        Error::new(ErrorKind::Other,
+                            format!("Could not forward message from client to server {:?}", e))
+                    )
+                )
+                .map(|(_from_client, _sink_err)| ())
+                .map_err(|_|
+                    Error::new(ErrorKind::Other,
+                        format!("Reader ended with error"))
+                );
+
+            processor
+                .select(reader)
+                    .map_err(|(err, _select_next)| {
+                        err
+                    })
+                    .map(|_| ())
+                .select(writer)
+                    .map_err(|(err, _select_next)| {
+                        err
+                    })
+        })
+        .map(|_| ());
+
+        Box::new(process)
+    });
+
+    Box::new(future)
+}
+
+/// Onion sink and stream for TCP.
+struct TcpOnion {
+    /// Sink for onion packets from TCP to UDP.
+    tx: mpsc::UnboundedSender<(OnionRequest, SocketAddr)>,
+    /// Stream of onion packets from TCP to UDP.
+    rx: mpsc::UnboundedReceiver<(InnerOnionResponse, SocketAddr)>,
+}
+
+/// Onion sink and stream for UDP.
+struct UdpOnion {
+    /// Sink for onion packets from UDP to TCP.
+    tx: mpsc::UnboundedSender<(InnerOnionResponse, SocketAddr)>,
+    /// Stream of onion packets from TCP to UDP.
+    rx: mpsc::UnboundedReceiver<(OnionRequest, SocketAddr)>,
+}
+
+/// Create onion streams for TCP and UDP servers communication.
+fn create_onion_streams() -> (TcpOnion, UdpOnion) {
+    let (udp_onion_tx, udp_onion_rx) = mpsc::unbounded();
+    let (tcp_onion_tx, tcp_onion_rx) = mpsc::unbounded();
+    let tcp_onion = TcpOnion {
+        tx: tcp_onion_tx,
+        rx: udp_onion_rx,
+    };
+    let udp_onion = UdpOnion {
+        tx: udp_onion_tx,
+        rx: tcp_onion_rx,
+    };
+    (tcp_onion, udp_onion)
+}
+
+fn run_tcp(cli_config: &CliConfig, dht_sk: SecretKey, tcp_onion: TcpOnion) -> impl Future<Item = (), Error = Error> {
+    let tcp_server = TcpServer::new_with_onion(tcp_onion.tx);
```

Comment:
Maybe add setter for onion as it was done for dht - `set_tcp_onion_sink`?

Reply:
Yes, but this will require another pr in https://github.com/tox-rs/tox-rs project.

Reply:
Do it please.

Reply:
https://github.com/tox-rs/tox/pull/173 merged